packages/flutter/lib/src/cupertino/dialog.dart

```diff
@@ -288,48 +277,192 @@ class CupertinoDialogAction extends StatelessWidget {
   }
 }
 
-const double _kButtonBarHeight = 45.0;
+/// Constructs a text content section typically used in a [CupertinoAlertDialog].
+///
+/// If [title] is missing, then only [message] is added.  If [message] is
+/// missing, then only [title] is added. If both are missing, then it returns
+/// null.
+class _CupertinoAlertTitleSection extends StatelessWidget {
+  const _CupertinoAlertTitleSection({
+    Key key,
+    this.title,
+    this.message,
+    this.scrollController,
+  }) : super(key: key);
 
-// This color isn't correct. Instead, we should carve a hole in the dialog and
-// show more of the background.
-const Color _kButtonDividerColor = const Color(0xFFD5D5D5);
+  /// The (optional) title of the dialog is displayed in a large font at the top
+  /// of the dialog.
+  ///
+  /// Typically a [Text] widget.
+  final Widget title;
 
-class _CupertinoButtonBar extends StatelessWidget {
-  const _CupertinoButtonBar({
+  /// The (optional) content of the dialog is displayed in the center of the
+  /// dialog in a lighter font.
+  ///
+  /// Typically a [Text] widget.
+  final Widget message;
+
+  /// A scroll controller that can be used to control the scrolling of the
+  /// message in the dialog.
+  ///
+  /// Defaults to null, and is typically not needed, since most alert messages
+  /// are short.
+  final ScrollController scrollController;
+
+  @override
+  Widget build(BuildContext context) {
+    final List<Widget> titleContentGroup = <Widget>[];
+    if (title != null) {
+      titleContentGroup.add(new Padding(
+        padding: new EdgeInsets.only(
+          left: _kEdgePadding,
+          right: _kEdgePadding,
+          bottom: message == null ? _kEdgePadding : 1.0,
+          top: _kEdgePadding,
+        ),
+        child: new DefaultTextStyle(
+          style: _kCupertinoDialogTitleStyle,
+          textAlign: TextAlign.center,
+          child: title,
+        ),
+      ));
+    }
+
+    if (message != null) {
+      titleContentGroup.add(
+        new Padding(
+          padding: new EdgeInsets.only(
+            left: _kEdgePadding,
+            right: _kEdgePadding,
+            bottom: _kEdgePadding,
+            top: title == null ? _kEdgePadding : 1.0,
+          ),
+          child: new DefaultTextStyle(
+            style: _kCupertinoDialogContentStyle,
+            textAlign: TextAlign.center,
+            child: message,
+          ),
+        ),
+      );
+    }
+
+    if (titleContentGroup.isEmpty) {
+      return new SingleChildScrollView(
+        controller: scrollController,
+        child: new Container(width: 0.0, height: 0.0),
+      );
+    }
+
+    // Add padding between the widgets if necessary.
+    if (titleContentGroup.length > 1) {
+      titleContentGroup.insert(1, const Padding(padding: const EdgeInsets.only(top: 8.0)));
+    }
+
+    return new Flexible(
+      flex: 3,
```

Comment:
why flex 3?

Reply:
generally, returning a Flexible is an anti-pattern, since it means the widget can only be used in a Flex. It would be better to move the Flexible into the place where you use this widget.